cmd/docker-driver/driver.go

```diff
@@ -0,0 +1,197 @@
+	}
+	if err := l.lokil.Close(); err != nil {
+		level.Error(l.logger).Log("msg", "error while closing loki logger", "err", err)
+	}
+	if err := l.jsonl.Close(); err != nil {
+		level.Error(l.logger).Log("msg", "error while closing json logger", "err", err)
+	}
+}
+
+func newDriver(logger log.Logger) *driver {
+	return &driver{
+		logs:   make(map[string]*logPair),
+		idx:    make(map[string]*logPair),
+		logger: logger,
+	}
+}
+
+func (d *driver) StartLogging(file string, logCtx logger.Info) error {
+	d.mu.Lock()
+	if _, exists := d.logs[file]; exists {
+		d.mu.Unlock()
+		return fmt.Errorf("logger for %q already exists", file)
+	}
+	d.mu.Unlock()
+
+	if logCtx.LogPath == "" {
+		logCtx.LogPath = filepath.Join("/var/log/docker", logCtx.ContainerID)
+	}
+	if err := os.MkdirAll(filepath.Dir(logCtx.LogPath), 0755); err != nil {
+		return errors.Wrap(err, "error setting up logger dir")
+	}
+	jsonl, err := jsonfilelog.New(logCtx)
+	if err != nil {
+		return errors.Wrap(err, "error creating jsonfile logger")
+	}
+
+	lokil, err := New(logCtx, d.logger)
+	if err != nil {
+		return errors.Wrap(err, "error creating loki logger")
+	}
+	level.Debug(d.logger).Log("msg", "Start logging", "id", logCtx.ContainerID, "file", file, "logpath", logCtx.LogPath)
+	f, err := fifo.OpenFifo(context.Background(), file, syscall.O_RDONLY, 0700)
+	if err != nil {
+		return errors.Wrapf(err, "error opening logger fifo: %q", file)
+	}
+
+	d.mu.Lock()
+	lf := &logPair{jsonl, lokil, f, logCtx, d.logger}
+	d.logs[file] = lf
+	d.idx[logCtx.ContainerID] = lf
+	d.mu.Unlock()
+
+	go consumeLog(lf)
+	return nil
+}
+
+func (d *driver) StopLogging(file string) {
+	level.Debug(d.logger).Log("msg", "Stop logging", "file", file)
+	d.mu.Lock()
+	lf, ok := d.logs[file]
+	if ok {
+		lf.Close()
+		delete(d.logs, file)
+	}
+	d.mu.Unlock()
+}
+
+func consumeLog(lf *logPair) {
+	dec := protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
+	defer dec.Close()
+	defer lf.Close()
+	var buf logdriver.LogEntry
+	for {
+		if err := dec.ReadMsg(&buf); err != nil {
+			if err == io.EOF || err == os.ErrClosed || strings.Contains(err.Error(), "file already closed") {
+				level.Debug(lf.logger).Log("msg", "shutting down log logger", "id", lf.info.ContainerID, "err", err)
+				return
+			}
+			dec = protoio.NewUint32DelimitedReader(lf.stream, binary.BigEndian, 1e6)
+		}
+		var msg logger.Message
+		msg.Line = buf.Line
+		msg.Source = buf.Source
+		if buf.PartialLogMetadata != nil {
+			msg.PLogMetaData.ID = buf.PartialLogMetadata.Id
+			msg.PLogMetaData.Last = buf.PartialLogMetadata.Last
+			msg.PLogMetaData.Ordinal = int(buf.PartialLogMetadata.Ordinal)
+		}
+		msg.Timestamp = time.Unix(0, buf.TimeNano)
+
+		// loki goes first as the json logger reset the message on completion.
+		if err := lf.lokil.Log(&msg); err != nil {
+			level.Error(lf.logger).Log("msg", "error pushing message to loki", "id", lf.info.ContainerID, "err", err, "message", msg)
+			continue
```

Comment:
Do we want to also not write logs to the file if we fail to push them to loki?

Reply:
yeah I think so, the driver example I used was skipping, but I think you're right.